Fix problem with recent Starcounter XSON

eriksunsol opened the pull request:
Recent versions of Starcounter XSON does not allow `Add` operations in JSON Patch sent from client side.

When `<juicy-select>` is dynamically inserted in a DOM based on view model structure, and the structure changes rapidly, the view model may change so that the element may be ripped out of the DOM before the `setTimeout()` on line 47 times out. When `value` is set on line 55, the view model property bound to the `value` attribute no longer exists in the view model, which triggers a JSON Patch to be sent to the server with an `Add` operation. Fail!

The proposed change protects the code in `setTimeout()` by ensuring that he element is still attached before doing anything potentially bad.

eriksunsol committed:
Fix problem with recent Starcounter XSON

Recent versions of Starcounter XSON does not allow `Add` operations in JSON Patch sent from client side.

When `<juicy-select>` is dynamically inserted in a DOM based on view model structure, and the structure changes rapidly, the view model may change so that the element may be ripped out of the DOM before the `setTimeout()` on line 47 times out. When `value` is set on line 55, the view model property bound to the `value` attribute no longer exists in the view model, which triggers a JSON Patch to be sent to the server with an `Add` operation. Fail!

The proposed change protects the code in `setTimeout()` by ensuring that he element is still attached before doing anything potentially bad.

eriksunsol commented:
Even better if the `setTimeout()` could be avoided altogether (bad practice in my mind, and most of the time a kludge to cover up something that is not right on a deeper level). I tried removing the `setTimeout()` and things seem to work properly still, but maybe I'm missing something.

warpech commented:
@eriksunsol thank you so much for isolating the problem. I have tried another approach by rewriting `juicy-select` to be more sync. Could you please check if this also solves the problem: #10

warpech reviewed (commented):
@eriksunsol thank you so much for isolating the problem. I have tried another approach by rewriting `juicy-select` to be more sync. 

Could you please check if this also solves the problem: PR #10?

miyconst commented on juicy-select.html:
```diff
@@ -42,17 +42,25 @@
                 this.$select.addEventListener("change", this.selectChange.bind(this));
                 this.set("multiple", this.$select.hasAttribute("multiple"));
 
+                this.isAttached = true;
```
As far as I know, the `isAttached` is now used by Polymer itself, so element can't use it for it's own purpose.

eriksunsol replied:
How about we use `this._isAttached` instead so we don't clash with Polymer?

miyconst commented on juicy-select.html:
```diff
@@ -42,17 +42,25 @@
+                this.isAttached = true;
+
                 setTimeout(function () {
+                    if (!this.isAttached)
```
What is the point of this check, if the property was set to `true` a few lines above? Is it because the element may get detached from the DOM during those 10 ms of timeout?

eriksunsol replied:
Yes, that's exactly why! It's not just in theory. It actually happened!

miyconst replied:
Then it's a bad way to solve this situation. Instead it should be:

```js
this._initializationTimeout = setTimeout(function () { ... });
```

```js
detached: function () {
    clearTimeout(this._initializationTimeout);
}
```

eriksunsol replied:
Obviously! 👍

eriksunsol committed:
Call clearTimeout() in detached() instead of updating flag.

As per [this comment](https://github.com/Juicy/juicy-select/pull/9/files/60b670aa68358fb55c399b707372901c782f93bc#r148023187) by @miyconst

warpech commented:
@eriksunsol PR #10 has an alternative solution that simplifies the data flow in `juicy-select`. Could you please check if it fixes the problem for you?

miyconst commented:
@warpech please feel free to decide which of those PR will end up in `master` if I am too late to answer.

eriksunsol commented:
@warpech I don't know if PR #10 fixes this problem as it introduces another problem instead. But since #10 is freed from all `setTimeout()` uses, the asynch problem I had should be gone.

The problem I'm facing is that all my `<juicy-select>` are now empty using PR #10. It's probably because our use of `<juicy-select>` is how @miyconst described it:

```HTML
<juicy-select>
    <select class="custom-class" on-change="onSelectChangeCustom"></select>
</juicy-select>
```

From your example I thought I could fix it by adding `slot="select"` to the inner `<select>` element, but it does not work. Maybe if you have some advice what needs to be changed where `<juicy-select>` is used I can give it another go.

eriksunsol commented:
@warpech I just fooled around a bit with PR #10 even though my select lists are empty, and it does __not__ solve this problem.

It seems that there is some asynchronicity in Polymer that surfaces instead.

Below is the stack trace leading up to palindrom generating an 'add' patch. Interesting points are:

1. `<juicy-select>`'s `optionsChanged` observer gets called from Polymer code and triggers the call that leads up to the `add` patch here: https://github.com/Juicy/juicy-select/blob/bdf3859afe6bd661843cd65efc437a111aef2925/juicy-select.html#L52
2. Polymer uses de-bouncing which introduces an asynchronous behavior.

The problem is thus just moved elsewhere. :-(

```
o (palindrom-dom.min.js:2)
set (palindrom-dom.min.js:2)
set (polymer.html:2212)
(anonymous) (polymer.html:1610)
_notifyListener (polymer.html:2123)
(anonymous) (polymer.html:1628)
fire (polymer.html:1337)
_notifyChange (polymer.html:1468)
_notifyEffect (polymer.html:1647)
_effectEffects (polymer.html:1501)
_propertySetter (polymer.html:1485)
setter (polymer.html:1564)
set (polymer.html:2217)
optionsChanged (juicy-select.html:52)         <---------------------------- 1.
_observerEffect (polymer.html:1656)
_effectEffects (polymer.html:1501)
_propertySetter (polymer.html:1485)
__setProperty (polymer.html:1494)
_applyEffectValue (polymer.html:1977)
_annotationEffect (polymer.html:1640)
_effectEffects (polymer.html:1501)
_propertySetter (polymer.html:1485)
setter (polymer.html:1564)
__setProperty (polymer.html:1496)
_applyEffectValue (polymer.html:1977)
_annotationEffect (polymer.html:1640)
_effectEffects (polymer.html:1501)
_propertySetter (polymer.html:1485)
setter (polymer.html:1564)
__setProperty (polymer.html:1496)
__setPropertyImpl (polymer.html:4312)
_applyEffectValue (polymer.html:1977)
_annotationEffect (polymer.html:1640)
_effectEffects (polymer.html:1501)
_propertySetter (polymer.html:1485)
__setProperty (polymer.html:1494)
__setPropertyImpl (polymer.html:4312)
_insertInstance (polymer.html:5034)
_render (polymer.html:4853)
complete (polymer-mini.html:2081)
Debouncer.boundComplete (polymer-mini.html:2058)   <---------------------- 2.
_atEndOfMicrotask (polymer-mini.html:2036)
window.MutationObserver.observe.characterData (polymer-mini.html:2051)
characterData (async)           <----------------------------- 2.
run (polymer-mini.html:2012)
go (polymer-mini.html:2067)
debounce (polymer-mini.html:2091)     <------------------------------------ 2.
debounce (polymer-mini.html:2100)      <------------------------------------ 2.
_debounceTemplate (polymer.html:4315)
_itemsChanged (polymer.html:4802)
_complexObserverEffect (polymer.html:1666)
_complexObserverPathEffect (polymer.html:2275)
_pathEffector (polymer.html:2254)
_pathEffectorImpl (polymer.html:4446)
template._pathEffector (polymer.html:4386)
_notifyPath (polymer.html:2155)
_annotationPathEffect (polymer.html:2269)
_pathEffector (polymer.html:2254)
_notifyPath (polymer.html:2155)
_notifyNodes
ImportedTemplatePrototype._notifyPath
StarcounterIncludePrototype._notifyPath
_annotationPathEffect (polymer.html:2269)
_pathEffector (polymer.html:2254)
_notifyPath (polymer.html:2155)
_forwardParentPath (polymer.html:5312)
_pathEffectorImpl (polymer.html:4442)
template._pathEffector (polymer.html:4386)
_notifyPath (polymer.html:2155)
_annotationPathEffect (polymer.html:2269)
_pathEffector (polymer.html:2254)
_notifyPath (polymer.html:2155)
_forwardItemPath (polymer.html:5105)
_itemsChanged (polymer.html:4805)
_complexObserverEffect (polymer.html:1666)
_complexObserverPathEffect (polymer.html:2275)
_pathEffector (polymer.html:2254)
_pathEffectorImpl (polymer.html:4446)
template._pathEffector (polymer.html:4386)
_notifyPath (polymer.html:2155)
_annotationPathEffect (polymer.html:2269)
_pathEffector (polymer.html:2254)
_notifyPath (polymer.html:2155)
_forwardParentPath (polymer.html:5312)
_pathEffectorImpl (polymer.html:4442)
template._pathEffector (polymer.html:4386)
_notifyPath (polymer.html:2155)
_annotationPathEffect (polymer.html:2269)
_pathEffector (polymer.html:2254)
_notifyPath (polymer.html:2155)
_notifyNodes
ImportedTemplatePrototype._notifyPath
StarcounterIncludePrototype._notifyPath
_annotationPathEffect (polymer.html:2269)
_pathEffector (polymer.html:2254)
_notifyPath (polymer.html:2155)
_forwardItemPath (polymer.html:5105)
_itemsChanged (polymer.html:4805)
_complexObserverEffect (polymer.html:1666)
_complexObserverPathEffect (polymer.html:2275)
_pathEffector (polymer.html:2254)
_pathEffectorImpl (polymer.html:4446)
template._pathEffector (polymer.html:4386)
_notifyPath (polymer.html:2155)
_annotationPathEffect (polymer.html:2269)
_pathEffector (polymer.html:2254)
_notifyPath (polymer.html:2155)
_forwardParentPath (polymer.html:5312)
_pathEffectorImpl (polymer.html:4442)
template._pathEffector (polymer.html:4386)
_notifyPath (polymer.html:2155)
_annotationPathEffect (polymer.html:2269)
_pathEffector (polymer.html:2254)
_notifyPath (polymer.html:2155)
_notifyNodes
ImportedTemplatePrototype._notifyPath
StarcounterIncludePrototype._notifyPath
_annotationPathEffect (polymer.html:2269)
_pathEffector (polymer.html:2254)
_notifyPath (polymer.html:2155)
_notifySplices (polymer.html:2335)
notifySplices (polymer.html:2327)
notifyTemplateDomBind
propagateDomBind
onPatchApplied
E.validateAndApplySequence (palindrom-dom.min.js:2)
(anonymous) (palindrom-dom.min.js:2)
t.receive (palindrom-dom.min.js:2)
n.receive (palindrom-dom.min.js:2)
E.handleRemoteChange (palindrom-dom.min.js:2)
o._ws.onmessage (palindrom-dom.min.js:2)
```

eriksunsol commented:
@warpech I'm sorry to say, but the problem is actually much worse with PR #10 than without.

warpech commented:
I am very thankful for your help in nailing down the problem. What is not tested, cannot be maintained and fixed. Sadly, this project has no tests yet, but it has a demo page that does not work on `master` branch but works in `fix-async-problem` branch.

I have just added a demo of using `<juicy-select><select></select></juicy-select>` in `fix-async-problem`. It seems to work as expected. Could you please give me a hint what is your usage that breaks it? 

I can only accept PR #9 or #10 if I can reproduce the behavior that it is fixing.

warpech commented:
If I understand correctly, the problem does not manifest in the UI of `juicy-select` but in an outgoing patch generated by Palindrom? Then it will might be hard to test here. Please confirm.

eriksunsol commented:
> I have just added a demo of using `<juicy-select><select></select></juicy-select>` in `fix-async-problem`. It seems to work as expected. Could you please give me a hint what is your usage that breaks it?

I don't have enough insight to understand why it doesn't work, but this is one of our uses where we don't get any options in the list with the `fix-async-problem` version:

https://github.com/HRidge/PalmaServer/blob/master/SideEditor/wwwroot/SideEditor/viewmodels/ToolEditors/DpmEditorPage.html

Looks similar to your example page, but it does not work. Maybe it's due to different nesting of custom elements, and/or the fact that it's in a Starcounter partial page. I don't know.

Maybe it's best if we can have a look at it together.

eriksunsol commented:
> If I understand correctly, the problem does not manifest in the UI of juicy-select but in an outgoing patch generated by Palindrom? Then it will might be hard to test here. Please confirm.

Correct, but I can describe in what way it misbehaves, and for that you could write test code. Pseudo-code below:

```JS
var model = { options: [], value: "" };

// Dynamically insert <juicy-select> element in DOM here and bind like
// <juicy-select value="{{model.value}}" options="{{model.options}}">.
// Insertion can e.g. be through a Starcounter partial page ( <starcounter-include> ).

// Do something to model.options to trigger a change notification handler in juicy-select
// (not needed prior to PR #10 as the juicy-select code itself had a setTimeout() call).

delete model.value;
setTimeout(function () {
  // The change notification triggered above is debounced in Polymer if run with PR #10 applied,
  // which will cause model.value to be updated before we get here, or without either PR #9 or
  // #10, the setTimeout call in the attached callback will update model.value.
  if (typeof(model.value) === "undefined") {
    console.log("Pass!");
  } else {
    console.log("Fail!");
  }
}, 1000);
```

warpech commented:
@alshakero will take a look at it today to:

- investigate how to reproduce the problem that we are fixing
- check once again if `setTimeout` can be removed completely
- add the demo page improvements from PR #10

alshakero commented:
@eriksunsol Hi! Could you please copy and paste the HTML where you use `<juicy-select>`, that would make my life much easier.

eriksunsol commented:
Here's a test page that will showcase the timing problem. Hit the test button to run the test.

```HTML
<!DOCTYPE html>
<html>
<head>
    <title>juicy-select</title>
	<meta charset="utf-8" />
    <script type="text/javascript" src="../webcomponentsjs/webcomponents-lite.js"></script>
    <link rel="import" href="../polymer/polymer.html" />
    <link rel="import" href="juicy-select.html" />
</head>
<body>
    <dom-bind>
        <template is="dom-bind" id="root">
            <h3 class="lead"><code>juicy-select</code> - Demo timing problem</h3>
            <div>Re-insert removed view model property:</div>
            <div><button type="button" on-click="reInsertViewModelProperty">Test!</button></div>
            <div><template is="dom-if" if="{{model.ShowTemp}}" restamp>
                <juicy-select caption-text="No favorite country" value="{{model.TempValue}}" options="{{model.Countries}}" selected-property="Selected$" text-property="Name" value-property="Name"></juicy-select>
            </template></div>
            <div>{{model.PassFail}}</div>

        </template>
        <script>
        
            var template = document.querySelector('dom-bind');
            if (!Polymer.Element) {
                // Polymer v 1.0
                template  = document.querySelector('#root');
            }
            
            template.model = {
                Countries: [{ Name: "Sweden", Selected: false }, { Name: "China", Selected: false }, { Name: "India", Selected: false }],
                ShowTemp: false,
                PassFail: "Not tested yet."
            };
            template.reInsertViewModelProperty = function () {
                // Set ShowTemp to true. Will show juicy-select element. Also insert TempValue.
                // These could both be Palindrom responding to one "Replace" and one "Add" patch.
                template.set("model.ShowTemp", true);
                template.set("model.TempValue", "Sweden");

                // Trigger change notification on options to fail PR #10.
                setTimeout(function() {
                    // We need to run this on the next iteration of the event loop so that Polymer
                    // has a chance to attach the juicy-select element first.
                    template.set("model.Countries", []);
                }, 0);

                // Set ShowTemp back to false and remove TempValue from the model.
                // This could be Palindrom responding to one "Replace" and one "Remove" patch
                // arriving immediately after the previous one.
                setTimeout(function() {
                    // We need to run this on the next iteration of the event loop, or Polymer will detect
                    // that this is a no-change, and no notification will be fired.
                    template.set("model.ShowTemp", false);

                }, 0);
                delete template.model.TempValue;

                // Wait...
                template.set("model.PassFail", "Running test...");
                setTimeout(function () {
                    // Before we get here any async handlers triggered by template.set("model.TempValue", "Sweden")
                    // will have run already.
                    if (typeof(template.model.TempValue) === "undefined") {
                        // TempValue remains absent from model. This is fine!
                        template.set("model.PassFail", "PASS!")
                    } else {
                        // TempValue has been added to the model. In this case Palindrom will have sent
                        // an "Add" patch to the server. XSON will throw an exception.
                        template.set("model.PassFail", "FAIL!")
                    }
                    // Reset model for another round.
                    delete template.model.TempValue;
                }, 1000);
            };
        </script>
    </dom-bind>
</body>
</html>

```

Not that because this is a timing problem, it's really difficult to construct a valid test case. This page fails on juicy-select release 1.0.0 and on PR #10. It's a pass on PR #9. I had to experiment a bit with where `setTimout()` calls should be to actually trigger the behavior that I had observed earlier with release 1.0.0 and PR #10.

Note that this code is different from the production code that fails because here a _Polymer_ `<template is="dom-if">` wraps the `<juicy-select>` element, while in production code it is `<starcounter-include>` or `<template is="imported-template">`. They may all behave different with respect to timing and synchronicity.

eriksunsol commented:
Note also that it is possible to modify the test page posted above to fail PR #9 too with slightly different use of `setTimeout()`. It's just a matter of how far one is prepared to go. The best way to solve this once and for all is by assuming that code outside of the juicy-select code can show any kind of asynchronous behavior, and then carefully review the juicy-select code with that taken into account.

alshakero commented:
Hi @eriksunsol, I enhanced the element a lot and added some tests that include your issue, could you do me the favour of pulling from [here](https://github.com/Juicy/juicy-select/pull/10) and testing? Thanks in advance.

warpech commented:
Ping @eriksunsol, could you please give a look to the version published [here](https://github.com/Juicy/juicy-select/pull/10)? We appreciate your help in closing this issue.

eriksunsol commented:
I'll try to find some time to check it out. Ping me again in a few days if I don't!

alshakero commented:
@eriksunsol ping 😄.

eriksunsol commented:
I've been using the #10 version today and haven't observed any problems with it. Reviewing the code, `options` and `value` are updated in `selectChange` only, which seems reasonable. The only danger I can see with respect to the original problem here (patches being sent to add data from client when using Palindrom) is that `multiple` may be set in `attached()` callback. But there is no `setTimeout()` involved, so assuming that the callback gets called without any delays, this should be fine too.

There is of course the odd chance that a user triggers the `change` event in the wrong time:

1. Patch from server causes removal of view-model sub-tree and removal of element from DOM to be scheduled.
2. User interacts with selection.
3. Node is added back to view-model and add patch sent to server.

But that would be assuming that removal of the element from the DOM is not completed on the same iteration of the event loop as the effects of the incoming patch is processed. If that's not the case, then it's safe. I depends on how Palindrom and e.g. `<starcounter-include>` interact (or whatever the mechanism is that manipulates the DOM based on the view-model) .

I think the current #10 version will be fine.

alshakero commented:
@eriksunsol thank you so much for your feedback Erik! I truly appreciate it.

warpech commented:
Setting as “Finished” because the fix in PR #10 was:
- released (with release notes) https://github.com/Juicy/juicy-select/releases/tag/1.1.0
